src/org/omegat/core/data/ExternalTMFactory.java

```diff
@@ -118,26 +123,45 @@ public ExternalTMX load(Language sourceLang, Language targetLang) throws Excepti
             TMXReader2.LoadCallback loader = new TMXReader2.LoadCallback() {
                 public boolean onEntry(TMXReader2.ParsedTu tu, TMXReader2.ParsedTuv tuvSource,
                         TMXReader2.ParsedTuv tuvTarget, boolean isParagraphSegtype) {
+
                     if (tuvSource == null) {
                         return false;
                     }
 
+                    // Keep all the Tuvs matching at least the target language
+                    for (TMXReader2.ParsedTuv tuv : tu.tuvs) {
+                        String[] tuvLangParts = tuv.lang.split("\\W");
+                        if (!targetLang.getLanguageCode().equalsIgnoreCase(tuvLangParts[0])) {
+                            continue;
+                        }
+                        addTuv(tu, tuvSource, tuv, isParagraphSegtype);
+                    }
+
+                    // If there was no match, use the original tuvTarget
                     if (tuvTarget != null) {
-                        // add only target Tuv
-                        addTuv(tu, tuvSource, tuvTarget, isParagraphSegtype);
+                        if (entries.isEmpty()) {
+                            addTuv(tu, tuvSource, tuvTarget, isParagraphSegtype);
+                        }
                     } else {
-                        // add all non-source Tuv
+                        // add all matching Tuv from other languages
                         for (int i = 0; i < tu.tuvs.size(); i++) {
-                            if (tu.tuvs.get(i) != tuvSource) {
-                                addTuv(tu, tuvSource, tu.tuvs.get(i), isParagraphSegtype);
+                            ParsedTuv tuvTarget2 = tu.tuvs.get(i);
+                            if (tuvTarget2 != tuvSource) {
+                                addTuv(tu, tuvSource, tuvTarget2, isParagraphSegtype, true);
```

Comment:
We are now walking the tuvs a second time, but only if `tuvTarget` was null... I think this if-else and the loop above it can probably be combined in a more understandable way. Since `tuvTarget` should be in `tu.tuvs`, can't we just loop once and add each entry in the appropriate manner?

Reply:
The loop is now simplified. There's a first one during TMX parsing to determine the best match. Though not really needed, I left it there to avoid breaking the LoadCallback interface) and after that there's a second loop in ExternalTMFactory to keep or leave out matches according to the user preferences.